File: dataset_parser/coders/apca/decoder_apca.py
import sys
sys.path.append('../')

from coders.cols.decoder_cols import DecoderCols


class DecoderAPCA(DecoderCols):
    def __init__(self, input_path, input_filename, output_csv, params):
        super(DecoderAPCA, self).__init__(input_path, input_filename, output_csv)
        self.window_size_bit_length = params['max_window_size'].bit_length()

    def _decode_column(self):
        self.row_index, self.column = 0, []

        while self.row_index < self.data_rows_count:
            window_length, value = self._decode_window()
            window = [value] * window_length
            self.column.extend(window)
            self.row_index += window_length

        return self.column

    def _decode_window(self, _=None):
        window_length = self.input_file.read_int(self.window_size_bit_length)
        value = self._decode_value_raw(self.row_index, self.column_index)
        return [window_length, value]

File: cpp_project/src/coders/apca/decoder_apca.cpp

#include "decoder_apca.h"

#include "string_utils.h"

void DecoderAPCA::setCoderParams(int max_window_size_){
    max_window_size_bit_length = StringUtils::bitLength(max_window_size_);
}

std::vector<std::string> DecoderAPCA::decodeColumn(){
    std::vector<std::string> column;
    row_index = 0;
    while (row_index < data_rows_count){
        decodeWindow(column);
    }
    return column;
}

void DecoderAPCA::decodeWindow(std::vector<std::string> & column){
    int window_size = input_file.getInt(max_window_size_bit_length);
    std::string value = decodeValueRaw();
    for (int i=0; i < window_size; i++){
        column.push_back(value);
        row_index++;
    }
}
